### ISSUE TITLE
Easier vehicle upgrading

### ISSUE THREAD

**Issue by LeftofZen** [OpenLoco/OpenLoco#1401]
### Describe the solution you'd like
I don't know exactly what the final solution looks like, but I have thought about it considerably and whilst I have some ideas, none seem to perfect and all seem to require large amounts of coding/dev work, so I'd like to ask others what they think would be a good solution to this problem. I've listed some ideas below but if anyone has any great ideas, please add your comments and thoughts.

The main goal is to make it easier to upgrade vehicles (ie trains). For save games with a large number of trains, when new vehicle types become available, it is a very manual-labour-intensive process to go through and upgrade your trains as you need to do either of the following sets of steps:

**Manual carriage replacement**
1. Select train
2. Stop train
3. Wait for train to stop
4. Drag carriages from the 'vehicle details' window onto the bin icon
5. Add new vehicles from the 'build vehicles' window
6. Optionally, you need to place the train on the tracks again if it doesn't fit the signal block
7. Start train

**Delete and clone**
1. Create upgraded version of train on separate track
2. Select train to be upgraded
3. Stop train
4. Wait for train to stop
5. Delete train
6. Clone new train and place on tracks at location of old train
7. Start train

### Ideas for a solution
- Add a 'depot' like OpenTTD, as described in #108. Then you can set a 'blueprint' in the depot, and then you can send a train to the depot to be upgraded at it's earliest convenience (basically add a temporary stop on the depot). In this case, the logic for the depot is simply when a train comes in, and it has 'upgrade' set on it, replace that train with the blueprinted one. On implementation, upgrading a train would become as simple as clicking the 'send to depot to upgrade' button, and the train would handle the rest itself. This is very simple for the player, but it would require an extremely large set of code changes, not to mention adding in the controversial 'depot' feature, which would be game-changing in and of itself.
- Same as the above idea, but remove the depot from the equation, keeping the blueprint idea. Then the upgrade gameplay loop would be stopping the train, clicking the upgrade button, optionally placing the train on tracks if it was unplaced, then starting it again. This would require a new in-game window to set up and store/manipulate the upgrade blueprints, but this could become a powerful feature of the game if implemented well.
- Often you only want to upgrade trains on a certain line, so potentially a different approach to this could be better 'line' controls, so for example we could add a control/feature to stop all trains on the same line (ie connected/block piece of track, similar to placing electric wires/3rd rail). A second option on this new control would be to give all trains on the same line the same orders (ie shared orders), and a third option would be to set all current trains on the line to the currently-selected train. This would mean you could upgrade 1 train manually, then bulk upgrade/replace everything on the same line at once. The downside to this idea though is that this only works on maps with cleanly-defined lines/connected track pieces, and doesn't allow you to upgrade only some of the trains on the line.

### Describe alternatives you've considered
Whatever the solution is, I think it will be good to design it in such a manner that the functionality can be built up piece by piece in small commits/PRs, rather than one large PR. Also, whatever functionality is added should be able to be applied to all vehicles, not just trains, and should also work with custom objects/vehicles.

### Additional context
N/A

**Comment by LeftofZen** [comment]
It's definitely a large issue which is why I tagged it with 'epic', and to be clear, it's not something I intend to seriously start implementing in the near future, I just wanted to see what other people thought of the idea and if they have any better ideas.

I did mention #108 in my post :p and whilst the idea appeals to me subjectively, I certainly don't see any of us implementing such a feature in the near future, if at all, I thought I'd just mention it as something I'd thought about.

**Comment by luk3ZPL** [comment]
You may also add a poll on the discord and you will see the results. Changes should be easy for implement and should not add more complexity. For me "upgrade button" or "upgrade tab" will be sufficient (replace when stopped).
![2](https://user-images.githubusercontent.com/4522457/159745297-ee990acf-8b9f-4206-a256-4394a861a20f.png)

**Comment by shusaura85** [comment]
Maybe something similar to OpenTTD's engine replacement window  

<img width="1280" height="744" alt="Image" src="https://github.com/user-attachments/assets/92757c6b-3383-4b5b-b8e7-0f070dc267bb" />
  
  
On the left it shows only the engines in use, on the right it shows all compatible engines that can be used as replacement.  
This actual replacements are done when the train visits a depot.  
For OpenLoco this could be done when the train stops at the station and finishes unloading but before loading (maybe do it only on the first or last station in the route list)  
The same interface is used for buses/trucks/planes/ships

**Comment by duncanspumpkin** [comment]
I have an alternative idea. "Lines"

You can optionally create a line from an individual train or just a blank one. A line can consist of multiple trains but each train has the same orders. You can quickly clone/remove trains based on the existing trains. You can also define an upgrade train set which will progressively replace the train sets in the line when they are stationary and have no cargo. You could even give a unique colour to all trains in the line. It would work the same for trains, planes, automobiles and boats.

This would solve multiple issues in one: 

- easier upgrading
- easier sharing of orders
- improved overview of all trains on a route

Disadvantages:

- What it doesn't do is allow upgrading all trains that you own at once. It would only allow upgrading them a line at a time. But perhaps you could share preset train set with other lines to simplify upgrading. 
- It might be painful to setup on an existing save but it's not exactly any worse than the current upgrade issues. 


Would that work? Above would require a NSF but none of it should be too difficult to achieve.

**Comment by shusaura85** [comment]
Sounds similar to the "Shared Orders" from OpenTTD  
>  Shared orders allow you to share one set of orders between multiple vehicles so that when one vehicle's orders are updated, the other vehicles sharing the same orders are also updated. It doesn't matter which vehicle you update - all vehicles sharing those orders are changed.  
  
OpenTTD also has the option to group the vehicles by the shared options (showing all on 1 row) (both for station vehicle list as well as all vehicle list)  
It also allows having more than one type of engine for a shared order group witch is more realistic than having the same engine for all vehicles  

<img width="1402" height="581" alt="Image" src="https://github.com/user-attachments/assets/aa9ce6f9-392a-4dd8-8b00-085e5ac385d6" />

Engine replacement can be part of the "Lines" feature but they are different features that can work together

**Comment by LeftofZen** [comment]
I already made a shared orders proposal here:

https://github.com/OpenLoco/OpenLoco/issues/1440

I don't think the shared orders and the vehicle upgrading should depend on each other, they should be individual components they can be added together if needed to make the "lines" work

**Comment by duncanspumpkin** [comment]
What we don't want to do is end up with a mess of lots of features like openttd in lots of sub windows and menus. I think it makes sense to combine these two or more parts into one as you will likely have identical train sets when you have identical orders.

But yes they would ultimately be two parts that are implemented separately and then joined together.

**Comment by LeftofZen** [comment]
Yeah combining in the UI is fine, I just want to keep the code/implementation clean and modular

**Comment by auge8472** [comment]
I don't think, the shared orders should be mixed with the vehicle replacement. The screenshot, @shusaura85 showed in [his latest comment](https://github.com/OpenLoco/OpenLoco/issues/1401#issuecomment-3619700763) is the vehicles list of a station window in OpenTTD (if I'm not misinterpreting, what I see). What's more accurate for a comparision are the common vehicle list windows in OpenTTD.

There one can configure grouping and also the replacements. With groups one can collect vehicles for specific freight but also vehicles serving specific regions or sharing their orders or for whatever systematics. One can optionally restrict the replacement orders to specific groups and within the groups to specific vehicle types. Without using the grouping feature the replacement would affect all vehicles of a specified type. IMHO this is a good system, at least for OpenTTD with its depots.

Having many lines with separate replacement orders sounds similar to me. However, I usually only use a few groups in OpenTTD, but I have a lot more timetables. This makes it seem very time-consuming for me to control the replacements via timetables.
